modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java

```diff
@@ -0,0 +1,240 @@
+
+        if (lowerBound == null) {
+            it.seekToFirst();
+        } else {
+            it.seek(lowerBound);
+        }
+
+        return new RocksIteratorAdapter<>(it) {
+            @Override
+            protected ByteBuffer decodeEntry(byte[] key, byte[] value) {
+                return ByteBuffer.wrap(key).order(ByteOrder.BIG_ENDIAN);
+            }
+
+            @Override
+            public void close() throws Exception {
+                super.close();
+
+                IgniteUtils.closeAll(options, upperBoundSlice);
+            }
+        };
+    }
+
+    private IndexRow decodeRow(ByteBuffer bytes) {
+        assert bytes.getShort(0) == partitionStorage.partitionId();
+
+        var tuple = new BinaryTuple(descriptor.binaryTupleSchema(), binaryTupleSlice(bytes));
+
+        // RowId UUID is located at the last 16 bytes of the key
+        long mostSignificantBits = bytes.getLong(bytes.limit() - Long.BYTES * 2);
+        long leastSignificantBits = bytes.getLong(bytes.limit() - Long.BYTES);
+
+        var rowId = new RowId(partitionStorage.partitionId(), mostSignificantBits, leastSignificantBits);
+
+        return new IndexRowImpl(tuple, rowId);
+    }
+
+    private byte[] rocksPrefix(BinaryTuplePrefix prefix) {
+        return rocksPrefix(prefix, 0).array();
+    }
+
+    private ByteBuffer rocksPrefix(InternalTuple prefix, int extraLength) {
+        ByteBuffer keyBytes = prefix.byteBuffer();
+
+        return ByteBuffer.allocate(Short.BYTES + keyBytes.remaining() + extraLength)
+                .order(ByteOrder.BIG_ENDIAN)
+                .putShort((short) partitionStorage.partitionId())
+                .put(keyBytes);
+    }
+
+    private byte[] rocksKey(IndexRow row) {
+        RowId rowId = row.rowId();
+
+        // We don't store the Partition ID as it is already a part of the key.
+        return rocksPrefix(row.indexColumns(), 2 * Long.BYTES)
+                .putLong(rowId.mostSignificantBits())
+                .putLong(rowId.leastSignificantBits())
+                .array();
+    }
+
+    private Predicate<ByteBuffer> startsWith(BinaryTuplePrefix prefix) {
+        var comparator = new BinaryTupleComparator(descriptor);
+
+        return key -> {
+            // First, compare the partitionIDs.
+            boolean partitionIdCompare = key.getShort(0) == partitionStorage.partitionId();
+
+            if (!partitionIdCompare) {
+                return false;
+            }
+
+            // Finally, compare the remaining parts of the tuples.
+            // TODO: This part may be optimized by comparing binary tuple representations directly. However, currently BinaryTuple prefixes
+            //  are not binary compatible with regular tuples. See https://issues.apache.org/jira/browse/IGNITE-17711.
+            return comparator.compare(prefix.byteBuffer(), binaryTupleSlice(key)) == 0;
+        };
+    }
+
+    private static ByteBuffer binaryTupleSlice(ByteBuffer key) {
+        return key.duplicate()
```

Comment:
Why duplicate instead of just slice?

Reply:
In order to create a slice I need to change the original buffer's position and limit. Do you have a better approach in mind?

Reply:
I thought ByteBuffer#slice(start, length) exists. But it is available only in java13+ :(